src/pgwire/src/protocol.rs

```diff
@@ -1386,6 +1401,97 @@ where
         Ok(State::Ready)
     }
 
+    async fn read_rows(
+        &mut self,
+        id: GlobalId,
+        columns: Vec<usize>,
+        params: CopyParams,
+        row_desc: RelationDesc,
+    ) -> Result<State, io::Error> {
+        let encode_format: pgrepr::Format = match params.format {
+            CopyFormat::Text => pgrepr::Format::Text,
+            // CopyFormat::Binary => pgrepr::Format::Binary,
+            _ => {
+                return self
+                    .error(ErrorResponse::error(
+                        SqlState::FEATURE_NOT_SUPPORTED,
+                        format!("COPY FROM format {:?} not supported", params.format),
+                    ))
+                    .await
+            }
+        };
+
+        let typ = row_desc.typ();
+        let column_formats = iter::repeat(encode_format)
+            .take(typ.column_types.len())
+            .collect::<Vec<pgrepr::Format>>();
+        self.conn
+            .send(BackendMessage::CopyInResponse {
+                overall_format: encode_format,
+                column_formats: column_formats.clone(),
+            })
+            .await?;
+        self.conn.flush().await?;
+
+        let mut data = Vec::new();
+        let mut next_state = State::Ready;
+        loop {
+            let message = self.conn.recv().await?;
+            match message {
+                Some(FrontendMessage::CopyData(buf)) => data.extend(buf),
+                Some(FrontendMessage::CopyDone) => break,
+                Some(_) => {
```

Comment:
The docs say:

> The backend will ignore Flush and Sync messages received during copy-in mode.

So those message types should be handled here.

Reply:
Thanks! Good catch!

I haven't added a test for Flush since the underlying library the test uses doesn't support it.